### ISSUE TITLE
Unexpected Side Effect from `ZkTracer.tracePreExecution`

### ISSUE THREAD

**Issue by DavePearce** [Consensys/besu-sequencer-plugins#949]
(see also #877 

Currently, we have this observation:

* [Block 6110045](https://lineascan.build/tx/0xc8574d2e389a6474f06e635870bb1a3b2c5fe73f515297733545f305861d4428#internal), second tx (`0xc8574d...`).  In the last `STATICCALL` gas goes from `238643` beforehand to `3572` within (where should be `234,816` within).

However, commenting out this code from `ZkTracer.tracePreExecution()` yields the expected gas usage:

```
if (frame.getCode().getSize() > 0) {
      try {
        this.hub.tracePreOpcode(frame);
        this.pin55.ifPresent(x -> x.tracePreOpcode(frame));
      } catch (final Exception e) {
        this.tracingExceptions.add(e);
      }
    }
```

Therefore, something somewhere is having an unexpected side effect on the EVM state.

**Comment by DavePearce** [comment]
Looks like `this.hub.tracePreOpcode(frame);` specifically is the problem.

**Comment by DavePearce** [comment]
In `Hub.java` this is where its coming from:

```
void processStateExec(MessageFrame frame) {
    ...
    this.handleStack(frame);
    this.triggerModules(frame);  <------
```

**Comment by DavePearce** [comment]
And then here:

```
void triggerModules(MessageFrame frame) {
   ...
    if (this.pch.signals().oob()) {
      this.oob.tracePreOpcode(frame);  <----
    }
```

**Comment by DavePearce** [comment]
Now here:

```
 public void tracePreOpcode(MessageFrame frame) { // This will be renamed to tracePreOp
    this.chunks.add(new OobOperation(frame, add, mod, wcp, hub, false, 0, 0));
    OpCode opCode = OpCode.of(frame.getCurrentOperation().getOpcode());

    if (opCode.isCall()) {
      Address target = Words.toAddress(frame.getStackItem(1));

     ...
        } else {
          // Other precompiles case
          this.chunks.add(new OobOperation(frame, add, mod, wcp, hub, true, 0, 0)); <---
        }
```

**Comment by DavePearce** [comment]
And now narrowed to somewhere in here:

```
private void populateColumns(final MessageFrame frame) { 
   ...
   } else if (isPrecompile()) {
      ... <---
   }
}
```

**Comment by DavePearce** [comment]
Ok, it seems to be this assignment which is causing the problem:

```
      final BigInteger callGas =
          BigInteger.valueOf(
              ZkTracer.gasCalculator.gasAvailableForChildCall(
                  frame, Words.clampedToLong(frame.getStackItem(0)), transfersValue));
```
Unsure why that would be.

**Comment by DavePearce** [comment]
Ok, looking in [TangerineWhistleGasCalculator.java](https://github.com/hyperledger/besu/blob/main/evm/src/main/java/org/hyperledger/besu/evm/gascalculator/TangerineWhistleGasCalculator.java) I find this:

```
@Override
  public long gasAvailableForChildCall(
      final MessageFrame frame, final long stipend, final boolean transfersValue) {
    final long gasCap = gasCap(frame.getRemainingGas(), stipend);

    // TODO: Integrate this into AbstractCallOperation since it's
    // a little out of place to mutate the frame here.
    frame.decrementRemainingGas(gasCap);

    if (transfersValue) {
      return gasCap + getAdditionalCallStipend();
    } else {
      return gasCap;
    }
  }
```

I believe this is the active implementation of `gasAvailableForChildCall` in `LondonGasCalculator`.  This is clearly *not* a pure function (which is surprising given its name).  See comment in the code!!

For the record, the function `gasAvailableForChildCall()` is being called from `OobOperation.populateColumns()`.  This could be resolved by manually calling `frame.incrementRemainingGas()` afterwards (which would require calculating the `gasCap()`).

@letypequividelespoubelles @powerslider FYI

**Comment by powerslider** [comment]
I am looking into this.

**Comment by OlivierBBB** [comment]
This is precisely what I fixed a few weeks ago on our HUB branch @DavePearce 🙂

See commit dc3a4fba869bc044e73b69118ff996acc51609df

**Comment by OlivierBBB** [comment]
The solution is not to re-increment but to extract the code. Or to ask Besu to write a side-effect free version of the method.

**Comment by letypequividelespoubelles** [comment]
see https://github.com/Consensys/linea-tracer/pull/748/commits/dc3a4fba869bc044e73b69118ff996acc51609df#diff-e35d4a0876dd8c77582a581789e424b9bf3bd60ef39555b43373db1e2b01c88d

**Comment by powerslider** [comment]
Thanks, I will try to adapt this to make it work.

**Comment by DavePearce** [comment]
> The solution is not to re-increment but to extract the code. Or to ask Besimu to write a side-effect free version of the method.

That would be the nicest solution for sure!!

**Comment by DavePearce** [comment]
Fixed with #950
